src/beast/beast/wsproto/impl/read_op.ipp

```diff
@@ -0,0 +1,161 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of Beast: https://github.com/vinniefalco/Beast
+    Copyright 2013, Vinnie Falco <vinnie.falco@gmail.com>
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef BEAST_WSPROTO_READ_OP_H_INCLUDED
+#define BEAST_WSPROTO_READ_OP_H_INCLUDED
+
+#include <beast/asio/handler_alloc.h>
+#include <memory>
+
+namespace beast {
+namespace wsproto {
+namespace detail {
+
+// read an entire message
+//
+template<class Stream, class Streambuf, class Handler>
+class read_op
```

Comment:
All the op classes have a similar structure. Here's an attempt to factor out that commonality: https://github.com/seelabs/rippled/commit/ee28b820b384805f6140e095a680316f61b2208b I only did that for the `read_op` but I think the other `op` classes could use `op_imp` too. What do you think? I don't like that there's extra types like `read_op_data` that shouldn't be instantiated on their own, but I do like removing some of the boilerplate.